xml/templates/dao.tpl

```diff
@@ -90,7 +95,7 @@ class {$table.className} extends CRM_Core_DAO {ldelim}
                  'name'      => '{$field.name}',
                                                                       'type'      => {$field.crmType},
 {if $field.title}
-                                                                      'title'     => ts('{$field.title}'),
+                                                                      'title'     => {if $extension}E::ts{else}ts{/if}('{$field.title}'),
```

Comment:
On a functional level, :+1: for having any DAO's in the extension use the extension's translation domain.

FWIW, when writing the template for the extension utils class, I explicitly wanted to avoid making `CRM_Foobar_ExtensionUtil` classes into a *public contract*. For example, it does not extend any shared base `class` or shared `interface`. The original/intended audience was only the author of the civix-generated module -- it was designed to be shared/public-contract.

Why?

(1) There are already public contracts for things like `ts(...'domain' => ...)` -- using the `E` class an inter-module contract would be duplicative and (2) civix is only supposed to be a template/skeleton generator. If an author wants to totally muck with all the code in there, then that should still be workable. But if others depend on civix-isms, then the dependency graph can get weird.

This patch is interesting because it lets the downstream inject the statement like `use CRM_Foobar_ExtensionUtil as E` -- technically, the core code doesn't know anything about `CRM_Foobar_ExtensionUtil`, but it does know about this virtual symbol `E` and `E::ts()`.

Would this patch mean that core code has some dependency on a civix convention? Honestly, it's right on the line. I think it's technically, currently on the safe side of the line. My hesitation is more with the precedent -- if other things in core start to depend on having an `E` class, then `E` will become a public contract leading eventually to funny loops.

Just to put it out there, here are a couple alternatives:

* (a) Have the civix generator pass the extension-key `org.example.foobar` or translation-domain for use in `dao.tpl`. In `dao.tpl`, it conditionally passes this for `ts(... ['domain' => ...])`.
* (b) Have the civix generator pass an expression for the translation callback (e.g. `$ts = 'CRM_Foobar_Extension::ts'`). In `dao.tpl`, it uses this instead of a straight call to `ts()`.

Reply:
@totten yeh i see your point but also I think this also about ensuring we have a separation between Core (ts) and the Extensions routing through their E utils class

Reply:
OK, so a routing point is useful because it gives you some leverage -- you make a change in one spot, and it applies to several use-cases.

`E::ts()` doesn't give us that:

1. `E::ts()` is used by newer PHP code, but it's not used by Smarty code, Javascript code, or older PHP code. To make `E::ts()` useful as a router, one would have to rework all those other call-paths.
2.  If you do make a change in `E::ts()`, it's liable to get trounced later on -- because it's part of the autogenerated `*.civix.php` boilerplate.

The common-denominator is that all paths lead to the global `ts()` function (with a `domain` argument). If one wants the benefit of routing (i.e. a leverage point where you can structurally manipulate all translations of a given extension/domain/purpose), then you currently have to look at the global `ts()`.

Reply:
@totten so then would you suggest that its better to perhaps add an extra function called like ts to extension DAOs that then calls the global ts function with the domain param?